### PR TITLE
Request Object for OIDC, initial refactoring to add JWT implementation

### PR DESCRIPTION
**Pull request by ayshsandu**
### Proposed changes in this pull request
- Refactoring code to accept JWT/JWE
- Necessary JWT validations
- Request Object Parameter validations
- Request Object Signature Validation with key alias
- Adding test cases

### When should this PR be merged

- Wait till parallel implementation by peers complete

### Follow up actions

- Test Request Object By Value flow with IS-5.5.x

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 1]
```diff
@@ -1123,12 +1137,15 @@ private String populateOauthParameters(OAuth2Parameters params, OAuthMessage oAu
             OIDC Request object will supersede parameters sent in the OAuth Authorization request. So handling the
             OIDC Request object needs to done after processing all request parameters.
          */
-        try {
-            handleOIDCRequestObject(oauthRequest, params);
-        } catch (RequestObjectException e) {
-            // All the error logs are specified at the time when throw the exception.
-            return EndpointUtil.getErrorPageURL(e.getErrorCode(), e.getErrorMessage(), null);
+        if (oauthRequest.getScopes().contains(OAuthConstants.Scope.OPENID)) {
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Can scopes be null ? If so let's do a null check

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java [hunk 2]
```diff
@@ -1953,23 +1953,15 @@ public static boolean isValidJson(String redirectURL) {
...
      */
-    public static List<String> essentialClaimsFromRequestParam(String claimRequestor, Map<String, List<Claim>>
+    public static List<String> essentialClaimsFromRequestParam(String claimRequestor, Map<String, List<RequestedClaim>>
             requestedClaimsFromRequestParam) {
 
-        String attributeValue = null;
         List<String> essentialClaimsfromRequestParam = new ArrayList<>();
-        List<Claim> claimsforClaimRequestor = requestedClaimsFromRequestParam.get(claimRequestor);
-        for (Claim claimforClaimRequestor : claimsforClaimRequestor) {
+        List<RequestedClaim> claimsforClaimRequestor = requestedClaimsFromRequestParam.get(claimRequestor);
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java]: can claimsforClaimRequestor be null ? If so don't we need to handle it ?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/OIDCRequestObjectFactory.java [hunk 3]
```diff
@@ -54,31 +57,49 @@ public static void buildRequestObject(OAuthAuthzRequest oauthRequest,
          */
         RequestObjectBuilder requestObjectBuilder;
         if (isRequestParameter(oauthRequest)) {
-             requestObjectBuilder = getRequestObjectBuilder(REQUEST_PARAM_VALUE_BUILDER);
-            if (requestObjectBuilder != null) {
-                requestObjectBuilder.buildRequestObject(oauthRequest.getParam(REQUEST), oAuth2Parameters, requestObject);
-                validateClientIdAndResponseType(oauthRequest, requestObject);
-            }
+            requestObjectBuilder = getRequestObjectBuilder(REQUEST_PARAM_VALUE_BUILDER);
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/OIDCRequestObjectFactory.java]: Is the way we have used builder pattern correct ?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/RequestObjectValidatorImpl.java [hunk 4]
```diff
@@ -62,110 +63,24 @@
...
-                if (log.isDebugEnabled()) {
-                    log.debug("As the request object is a nested jwt, passed the payload to validate the signature.");
-                }
-            }
-        } catch (JOSEException | IdentityOAuth2Exception | java.text.ParseException e) {
-            throw new RequestObjectException(RequestObjectException.ERROR_CODE_INVALID_REQUEST, "Failed to decrypt " +
-                    "request object.");
-        }
+    public boolean validateSignature(RequestObject requestObject, OAuth2Parameters oAuth2Parameters) throws
+            RequestObjectException {
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/RequestObjectValidatorImpl.java]: New line after method ?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/RequestObjectValidatorImpl.java [hunk 5]
```diff
@@ -175,177 +90,204 @@ private RSAPrivateKey getRsaPrivateKey(OAuth2Parameters oAuth2Parameters) throws
...
-
-        if (!OAuth2Util.isValidJson(requestObject)) {
-            String[] jwtTokenValues = requestObject.split("\\.");
-            if (jwtTokenValues.length == NUMBER_OF_PARTS_IN_JWS) {
-                processRequestObject(jwtTokenValues);
-                validateSignature(requestObject);
-            } else if (jwtTokenValues.length == NUMBER_OF_PARTS_IN_JWE) {
-                decrypt(requestObject, oAuth2Parameters);
-            }
+        boolean isValid = validateClientIdAndResponseType(requestObject, oAuth2Parameters);
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/RequestObjectValidatorImpl.java]: New line after method ?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/RequestObjectValidatorImpl.java [hunk 6]
```diff
@@ -175,177 +90,204 @@ private RSAPrivateKey getRsaPrivateKey(OAuth2Parameters oAuth2Parameters) throws
...
+                            IdentityApplicationConstants.Authenticator.OIDC.NAME);
+            tokenEndpoint = IdentityApplicationManagementUtil.getProperty(oidcFedAuthn.getProperties(),
+                    IdentityApplicationConstants.Authenticator.OIDC.OAUTH2_TOKEN_URL).getValue();
+            if (log.isDebugEnabled()) {
+                log.debug("Found Token Endpoint URL: " + tokenEndpoint);
             }
-        } else {
-            throw new RequestObjectException(RequestObjectException.ERROR_CODE_INVALID_REQUEST, "Signature is null.");
+        } catch (IdentityProviderManagementException e) {
+            log.error("Error while loading OAuth2TokenEPUrl of the resident IDP of tenant:" + tenantDomain, e);
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/RequestObjectValidatorImpl.java]: Why do we log and throw ?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/RequestObjectValidatorImpl.java [hunk 7]
```diff
@@ -175,177 +90,204 @@ private RSAPrivateKey getRsaPrivateKey(OAuth2Parameters oAuth2Parameters) throws
...
+            } else {
+                // for super tenant, load the default pub. cert using the config. in carbon.xml
+                keyStore = keyStoreManager.getPrimaryKeyStore();
             }
+            return keyStore.getCertificate(alias);
+
+        } catch (KeyStoreException e) {
+            String errorMsg = "Error instantiating an X509Certificate object for the certificate alias:" + alias +
+                    " in tenant:" + tenantDomain;
+            log.error(errorMsg, e);
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/RequestObjectValidatorImpl.java]: Do we need to do both logging and throwing ? Isn't throwing enough ?

**Comment by ayshsandu** [comment]
Closing since code is moved to new feature branch.

**Pull request closed without merging**